lisa/test_runner/tests/test_lisarunner.py

```diff
@@ -0,0 +1,411 @@
+        runner._merge_test_requirements(
+            test_results=test_results,
+            existing_environments=envs,
+            platform_type=constants.PLATFORM_MOCK,
+        )
+        self.assertListEqual(
+            [], [x for x in envs],
+        )
+        self.assertListEqual(
+            [TestStatus.SKIPPED, TestStatus.SKIPPED, TestStatus.SKIPPED],
+            [x.status for x in test_results],
+        )
+        not_allow_new_message = (
+            "not found fit environment, and not allow to create new environment"
+        )
+        self.assertListEqual(
+            [not_allow_new_message, not_allow_new_message, not_allow_new_message],
+            [x.message for x in test_results],
+        )
+
+    def test_merge_req_platform_type_checked(self) -> None:
+        # checked is enough. More covered in search space
+        env_runbook = generate_env_runbook(is_single_env=False)
+        envs = load_environments(env_runbook)
+        self.assertListEqual(
+            [], [x for x in envs],
+        )
+        runner = generate_lisarunner(None)
+        test_results = generate_cases_result()
+        for test_result in test_results:
+            metadata = test_result.runtime_data.metadata
+            metadata.requirement = simple_requirement(
+                supported_platform_type=["notexists"]
+            )
+        runner._merge_test_requirements(
+            test_results=test_results,
+            existing_environments=envs,
+            platform_type=constants.PLATFORM_MOCK,
+        )
+        self.assertListEqual(
+            [TestStatus.SKIPPED, TestStatus.SKIPPED, TestStatus.SKIPPED],
+            [x.status for x in test_results],
+        )
+        platform_unsupported = "capability cannot support some of requirement"
+        self.assertListEqual(
+            [platform_unsupported, platform_unsupported, platform_unsupported],
+            [x.message[0:45] for x in test_results],
+        )
+
+    def test_fit_a_predefined_env(self) -> None:
+        # with predefined env of 1 simple node, so ut2 don't need a new env
+        generate_cases_metadata()
+        env_runbook = generate_env_runbook(is_single_env=True, remote=True)
+        runner = generate_lisarunner(env_runbook)
+        asyncio.run(runner.start())
+        self.assertListEqual(
+            ["runbook_0", "req_1", "req_2"], [x for x in prepared_envs],
+        )
+        self.assertListEqual(
+            ["runbook_0", "req_1"], [x for x in deployed_envs],
+        )
+        self.assertListEqual(
+            ["runbook_0", "req_1"], [x for x in deleted_envs],
+        )
+        self.assertListEqual(
+            ["req_1", "runbook_0", "runbook_0"],
+            [x.assigned_env for x in runner._latest_test_results],
+        )
+        self.assertListEqual(
+            [TestStatus.PASSED, TestStatus.PASSED, TestStatus.PASSED],
+            [x.status for x in runner._latest_test_results],
+        )
+        self.assertListEqual(
```

Comment:
I noticed that a lot of the tests _could_ use iterator checks like `all()`; however, since this is unit test code, I actually am inclined to prefer the explicit list comparisons (how it's currently written).

Reply:
@lpereira what do you think? Being unit test code, I mostly appreciate that (a) it exists (b) it works and (c) it's very explicit.

Reply:
While I generally prefer to use predicates like `any()` or `all()`, I agree that, on unit-tests, copy-paste+explicit code is better.

Reply:
all or any will hide expected, and actual informations. If a test case failed, it has to be debugged to see actual values. Assert for list and dict are very helpful to see details when failed in lab.

I will refactor this file to make more code shared, and add some comments to specify what's tested.